src/binaryServerStarter.ts

```diff
@@ -0,0 +1,150 @@
+ */
+export async function prepareBinaryExecutable(context: ExtensionContext): Promise<Executable> {
+  const binaryOptions: string = getXMLConfiguration().get("server.binary.args");
+  let binaryExecutable: Executable;
+  return getServerBinaryPath()
+    .then((binaryPath: string) => {
+      binaryExecutable = {
+        args: [binaryOptions],
+        command: binaryPath
+      } as Executable;
+      return binaryPath;
+    })
+    .then(checkBinaryHash)
+    .then((hashOk: boolean) => {
+      if (hashOk) {
+        return binaryExecutable;
+      } else {
+        return new Promise((resolve, reject) => {
+          return window.showErrorMessage(`The server binary ${binaryExecutable.command} is not trusted. `
+            + 'Running the file poses a threat to your system\'s security. '
+            + 'Run anyways?', 'Yes', 'No')
+            .then((val: string) => {
+              if (val === 'Yes') {
+                resolve(binaryExecutable);
+              }
+              reject("The binary XML language server is not trusted");
+            });
+        });
+      }
+    });
+}
+
+/**
+ * Returns the path to the LemMinX binary
+ *
+ * Downloads it if it is missing
+ *
+ * @returns The path to the LemMinX binary
+ * @throws If the LemMinX binary can't be located or downloaded
+ */
+async function getServerBinaryPath(): Promise<string> {
+  const config: WorkspaceConfiguration = getXMLConfiguration();
+  let binaryPath: string = config.get("server.binary.path");
+  if (binaryPath) {
+    if (fs.existsSync(binaryPath)) {
+      return Promise.resolve(binaryPath);
+    }
+    window.showErrorMessage('The specified XML language server binary could not be found. Using the default binary...');
+  }
+  let server_home: string = path.resolve(__dirname, '../server');
+  let binaries: Array<string> = glob.sync(`**/lemminx-${os.platform()}*`, { cwd: server_home });
+  const JAR_AND_HASH_REJECTOR: RegExp = /(\.jar)|(hash)$/;
+  binaries = binaries.filter((path) => { return !JAR_AND_HASH_REJECTOR.test(path) });
+  if (binaries.length) {
+    return new Promise<string>((resolve, reject) => {
+      resolve(path.resolve(server_home, binaries[0]));
+    });
+  }
+  // Download it, then return the downloaded binary's location
+  return downloadBinary();
+}
+
+/**
+ * Downloads LemMinX binary under the `server` directory and returns the path to the binary as a Promise
+ *
+ * @returns The path to the LemMinX binary
+ * @throws If the LemMinX binary download fails
+ */
+async function downloadBinary(): Promise<string> {
+  window.setStatusBarMessage('Downloading XML language server binary...', 2000);
+  return new Promise((resolve, reject) => {
+    const handleResponse = (response: http.IncomingMessage) => {
+      const statusCode = response.statusCode;
+      if (statusCode === 303) {
+        http.get(response.headers.location, handleResponse);
+      } else if (statusCode === 200) {
```

Comment:
we'll most likely download zip files. If the MIME type is application/zip, then we should unzip it first

Reply:
using zip will save bandwidth, bigly ;-)
-rw-r--r--@ 1 fbricon  staff    12M Nov 12 10:06 lemminx-darwin.zip
-rw-r--r--@ 1 fbricon  staff    38M Nov  4 19:26 lemminx-osx-x86_64-0.14.0-SNAPSHOT

Reply:
It should work now. Needed to include yauzl to unzip files. The download links are now embedded in the package.json